### DISCUSSION TITLE
关于服务发现，是否有直接使用etcd的库

### DISCUSSION THREAD

**Discussion by attson** [apache/apisix#8742] [Q&A] [answered]
了解到apisix本身也是读取etcd的数据。

我认为一个注册中心最核心的能力
- keepalive
- subscribe
- meta
- availability
这些etcd本身就是具备的。

是否可以有一个 注册中心 sdk，直接与etcd交互，应用启动的时候，通过sdk上报当前应用的节点信息，路由信息。apisix照常读取 etcd的信息。

在上一家公司我们自研的一套注册中心和配置中心就是直接基于etcd的，如果apisix也支持这种方式接入，就可以不用使用 nacos 之类的其他中间件了。

不知道我的思路是否有问题。
如果说这样是可行的，那是否已经有相关的开源组件（避免造轮子）

**Comment by tokers** [comment, marked as answer]
It's OK to use ETCD as the service registry, but it lacks a standard for how to register some service instances and how to discover them. I think a proposal is required.

**Reply by attson** [reply]
ok, 我认为这是有价值的。

不过，我刚才看了一下 etcd upstream 中的结构

```
/apisix/routes/445492773978636991
{"id":"445492773978636991","create_time":1675063985,"update_time":1675065325,"uri":"/web1/*","name":"web1","methods":["GET","POST","PUT","DELETE","PATCH","HEAD","OPTIONS","CONNECT","TRACE","PURGE"],"service_id":"445494910154441407","status":1}

/apisix/upstreams/445492582533825215
{"id":"445492582533825215","create_time":1675063871,"update_time":1675064082,"nodes":{"web1:80":1},"timeout":{"connect":6,"send":6,"read":6},"type":"roundrobin","scheme":"http","pass_host":"pass","name":"web1","keepalive_pool":{"idle_timeout":60,"requests":1000,"size":320

/apisix/services/445494910154441407
{"id":"445494910154441407","create_time":1675065258,"update_time":1675065258,"name":"web1","upstream_id":"445492582533825215"}
```

routes: 如果一个服务对应一个路由组，用服务名作为 url 前缀，这个结构是没有多大问题的
upstreams: 一个服务对应一个 upstreams，对于node keepalive场景，通常是以node 维度单独一个key，与现有的结构是不太相符的。
services: 与 upstreams 在服务注册场景，其实保留一个就挺好
